transports/quic/src/transport.rs

```diff
@@ -149,45 +161,66 @@ impl<P: Provider> Transport for GenTransport<P> {
                     return false;
                 }
-                let listen_addr = l.endpoint_channel.socket_addr();
+                let listen_addr = l.socket_addr();
                 SocketFamily::is_same(&listen_addr.ip(), &socket_addr.ip())
                     && listen_addr.ip().is_loopback() == socket_addr.ip().is_loopback()
             })
             .collect::<Vec<_>>();
 
-        let dialer_state = match listeners.len() {
+        let endpoint = match listeners.len() {
             0 => {
                 // No listener. Get or create an explicit dialer.
                 let socket_family = socket_addr.ip().into();
                 let dialer = match self.dialer.entry(socket_family) {
-                    Entry::Occupied(occupied) => occupied.into_mut(),
+                    Entry::Occupied(occupied) => occupied.get().clone(),
                     Entry::Vacant(vacant) => {
                         if let Some(waker) = self.waker.take() {
                             waker.wake();
                         }
-                        vacant.insert(Dialer::new::<P>(self.quinn_config.clone(), socket_family)?)
+                        let listen_socket_addr = match socket_family {
+                            SocketFamily::Ipv4 => SocketAddr::new(Ipv4Addr::UNSPECIFIED.into(), 0),
+                            SocketFamily::Ipv6 => SocketAddr::new(Ipv6Addr::UNSPECIFIED.into(), 0),
+                        };
+                        let endpoint_config = self.quinn_config.endpoint_config.clone();
+                        let endpoint =
+                            Self::new_endpoint(endpoint_config, None, listen_socket_addr)?;
+
+                        vacant.insert(endpoint.clone());
+                        endpoint
                     }
                 };
-                &mut dialer.state
+                dialer
             }
-            1 => &mut listeners[0].dialer_state,
             _ => {
                 // Pick any listener to use for dialing.
                 // We hash the socket address to achieve determinism.
                 let mut hasher = DefaultHasher::new();
                 socket_addr.hash(&mut hasher);
                 let index = hasher.finish() as usize % listeners.len();
-                &mut listeners[index].dialer_state
+                listeners[index].endpoint.clone()
             }
         };
-        Ok(dialer_state.new_dial(socket_addr, self.handshake_timeout, version))
+        let handshake_timeout = self.handshake_timeout;
+        let mut client_config = self.quinn_config.client_config.clone();
+        if version == ProtocolVersion::Draft29 {
+            client_config.version(0xff00_001d);
+        }
+        Ok(Box::pin(async move {
+            // This `"l"` seems necessary because an empty string is an invalid domain
+            // name. While we don't use domain names, the underlying rustls library
+            // is based upon the assumption that we do.
```

Comment:
What does `l` mean and can we set something more self-descriptive like "wedontusedomainnames.com"? :D

Reply:
:D maybe it will work.
It's an artifact from the prev impl: https://github.com/libp2p/rust-libp2p/blob/57b6605e7128f6cd1d60921c4c4d51d196b62a67/transports/quic/src/endpoint.rs#L436